include/random_walks/exponential_hamiltonian_monte_carlo_exact_walk.hpp

```diff
@@ -0,0 +1,284 @@
+    {
+        _Len = L;
+    }
+
+
+private :
+
+    template
+    <
+        typename GenericPolytope
+    >
+    inline void initialize(GenericPolytope const& P,
+                           Point const& p,
+                           RandomNumberGenerator &rng)
+    {
+        unsigned int n = P.dimension();
+        NT T;
+        _lambdas.setZero(P.num_of_hyperplanes());
+        _Av.setZero(P.num_of_hyperplanes());
+        _v = GetDirection<Point>::apply(n, rng, false);
+        
+        do {
+            _p = p;
+            T = rng.sample_urdist()* _Len;
+            int it = 0;
+
+            std::pair<NT, int> pbpair
+                    = P.quadratic_positive_intersect(_p, _v, _Ac, _Temp, _lambdas, _Av, _facet_prev);
+            if (T <= pbpair.first || pbpair.second < 0) {
+                _p += ((T * T) / (-2.0*_Temp)) *_c + (T * _v);
+                _lambda_prev = T;
+                return;
+            }
+            _lambda_prev = pbpair.first;
+            _p += ((_lambda_prev * _lambda_prev) / (-2.0*_Temp)) *_c + (_lambda_prev * _v);
+            _v += (-_lambda_prev/_Temp) * _c;
+            T -= _lambda_prev;
+            P.compute_reflection(_v, _p, pbpair.second);
+
+            while (it <= 100*n)
+            {
+                std::pair<NT, int> pbpair
+                        = P.quadratic_positive_intersect(_p, _v, _Ac, _Temp, _lambdas, _Av, _lambda_prev, _facet_prev);
+                if (T <= pbpair.first || pbpair.second < 0) {
+                    _p += ((T * T) / (-2.0*_Temp)) *_c + (T * _v);
+                    _lambda_prev = T;
+                    break;
+                }else if (it == 100*n) {
+                    _lambda_prev = rng.sample_urdist() * pbpair.first;
+                    _p += ((_lambda_prev * _lambda_prev) / (-2.0*_Temp)) *_c + (_lambda_prev * _v);
+                    break;
+                }
+                _lambda_prev = pbpair.first;
+                _p += ((_lambda_prev * _lambda_prev) / (-2.0*_Temp)) *_c + (_lambda_prev * _v);
+                _v += (-_lambda_prev/_Temp) * _c;
+                T -= _lambda_prev;
+                P.compute_reflection(_v, _p, pbpair.second);
+                it++;
+            }
+        } while (P.is_in(_p, _tol) == 0);
+    }
+
+
+    inline double get_max_distance(std::vector<Point> &pointset, Point const& q, double &rad) 
+    {
+        double dis = -1.0, temp_dis;
+        int jj = 0;
+        for (auto vecit = pointset.begin(); vecit!=pointset.end(); vecit++, jj++) 
+        {
+            temp_dis = (q.getCoefficients() - (*vecit).getCoefficients()).norm();
+            if (temp_dis > dis) {
+                dis = temp_dis;
+            }
+            if (jj == 0) {
+                if (temp_dis > rad) {
+                    rad = temp_dis;
+                }
+            }
+        }
+        return dis;
+    }
+
+
+    NT _Len;
+    VT _Ac;
+    Point _p;
+    Point _v;
+    Point _c;
+    NT _Temp;
+    const double _tol = 1e-10;
```

Comment:
this could be `static`, right?

Reply:
Thanks for the comment. However, this will give us a compile error in the R interface. 
I used `#define IN_INSIDE_BODY_TOLLERANCE 1e-10`.